pkg/volume/csi/csi_attacher.go

```diff
@@ -262,10 +265,12 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 		if cleanErr := os.RemoveAll(dataDir); cleanErr != nil {
 			klog.Error(log("failed to remove dir after error [%s]: %v", dataDir, cleanErr))
 		}
-		return err
+		return opExitStatus, err
 	}
 	defer func() {
-		if err != nil {
+		// Only if there was an error and volume operation was considered
+		// finished, we should remove the directory.
+		if err != nil && opExitStatus == volumetypes.OperationFinished {
```

Comment:
if fail to get credentials, the state is OperationStateNoChange, I think it will not mark volume in the actual state. Then is there any code we will clean up metadata?

Reply:
fixed